src/garage/torch/algos/vpg.py

```diff
@@ -0,0 +1,230 @@
+        self.entropy_method = entropy_method
+        self.eps = 1e-8
+
+        self.maximum_entropy = (entropy_method == 'max')
+        self.entropy_regularzied = (entropy_method == 'regularized')
+        self._check_entropy_configuration(entropy_method, center_adv,
+                                          stop_entropy_gradient,
+                                          policy_ent_coeff)
+        self._episode_reward_mean = collections.deque(maxlen=100)
+        self.optimizer = optimizer(policy.parameters(), lr=policy_lr)
+
+        filter = torch.full((1, 1, 1, max_path_length), discount * gae_lambda)
+        self.filter = torch.cumprod(filter, dim=-1)
+
+        super().__init__(policy=policy,
+                         baseline=baseline,
+                         discount=discount,
+                         max_path_length=max_path_length,
+                         n_samples=n_samples)
+
+    def _check_entropy_configuration(self, entropy_method, center_adv,
+                                     stop_entropy_gradient, policy_ent_coeff):
+        if entropy_method not in ('max', 'regularized', 'no_entropy'):
+            raise ValueError('Invalid entropy_method')
+
+        if entropy_method == 'max':
+            if center_adv:
+                raise ValueError('center_adv should be False when '
+                                 'entropy_method is max')
+            if not stop_entropy_gradient:
+                raise ValueError('stop_gradient should be True when '
+                                 'entropy_method is max')
+        if entropy_method == 'no_entropy':
+            if policy_ent_coeff != 0.0:
+                raise ValueError('policy_ent_coeff should be zero '
+                                 'when there is no entropy method')
+
+    def train_once(self, itr, paths):
+        """Perform one step of policy optimization."""
+        samples_data = self.process_samples(itr, paths)
+
+        self._optimize_policy(itr, paths)
+        self.baseline.fit(paths)
+
+        return samples_data['average_return']
+
+    def _optimize_policy(self, itr, paths):
+        self.optimizer.zero_grad()
+
+        policy_entropies = torch.stack([
+            self._add_padding(
+                self.policy.get_entropy(torch.Tensor(path['observations'])),
+                self.max_path_length)  # noqa: E501
+            for path in paths
+        ])
+
+        valids = [len(path['actions']) for path in paths]
+
+        baselines = torch.stack([
+            self._add_padding(self._get_baselines(path), self.max_path_length)
+            for path in paths
+        ])
+
+        rewards = torch.stack([
+            self._add_padding(torch.Tensor(path['rewards']),
+                              self.max_path_length)  # noqa: E501
+            for path in paths
+        ])
+
+        if self.maximum_entropy:
+            rewards += self.policy_ent_coeff * policy_entropies
+
+        deltas = (rewards + self.discount * F.pad(baselines,
```

Comment:
there is a helper function for this calculation in TF, because doing it with convolution can be quite tricky.

can you pull the advantage calculation out into a helper function and write a unit test for it? https://github.com/rlworkgroup/garage/blob/master/src/garage/tf/misc/tensor_utils.py#L201

you can also use your helper function to verify that you get the same results as the TF version for the same inputs.

Reply:
extract `compute_advantages` function and make test case for the function